### PR TITLE
Add merge_and_return_zero_coin function on coin.move for gas optimize reason.

### PR DESCRIPTION
**Pull request by e99243506bigplay**
… reason.

### Description
Could we have this? so some dapps I think it could use this function to avoid destroy coin then create coin again for gas optimizing. Please let me know if there has something wrong.

### Test Plan

I wrote the spec for verify propose, it should be right, what test plan should I did?

--------
Honestly, if you want less gas fee on Move Contract, you have to write the contract like assembly language and exploit gas meter rule.

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/coin.move [hunk 1]
```diff
@@ -470,6 +470,18 @@ module aptos_framework::coin {
         let Coin { value: _ } = source_coin;
     }
 
+    /// Same as be merge function but it return zero source coin for gas fee optimize reason if
+    /// you got a situation that you need merge Coin<CoinType> then you also need zero value Coin<CoinType>,
+    /// for example to feed zero coin to an Uniswap V2 type AMM DEX for doing a swap action.
+    public fun merge_and_return_zero_coin<CoinType>(dst_coin: &mut Coin<CoinType>, source_coin: Coin<CoinType>) : Coin<CoinType> {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/coin.move:476]: Why not just pass a &mut Coin<CoinType> for source_coin?

Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/coin.move:476]: Can you add a unit test for this?

**Review by movekevin — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/coin.move (hunk 1 above)
Comment (e99243506bigplay) [reply aptos-move/framework/aptos-framework/sources/coin.move:476]: > Why not just pass a &mut Coin for source_coin?

That will add more move instruction like `ReadRef` and `WriteRef` in low level bytecode, by that method may will negate gas optimize effort.

**Comment by e99243506bigplay** [comment]
Let me writing a new test case for this propose...

**Comment by e99243506bigplay** [comment]
Now, I am going to decompile the module, to see what is it inside there.

**Comment by e99243506bigplay** [comment]
Somehow, I can't prove the module lol, I maybe should reinstall by `./script/dev_setup.sh` again.

**Comment by e99243506bigplay** [comment]
Wait... I realized now...

**Comment by e99243506bigplay** [comment]
I feel the merge has some space on gas optimize, if we could make merge function of bytecode instruction as be:

```
public merge<Ty0>(Arg0: &mut Coin<Ty0>, Arg1: Coin<Ty0>) {
B0:
    1: CopyLoc[0](Arg0: &mut Coin<Ty0>)
    2: ImmBorrowFieldGeneric[1](Coin.value: u64)
    3: ReadRef // stack now has [dst_coin.value]

    4: MoveLoc[1](Arg1: Coin<Ty0>) // input, stack should be [src_coin, dst_coin.value]
    5: UnpackGeneric[1](Coin<Ty0>) // unpack, stack should be [src_coin.value, dst_coin.value]

    6: Add // src_coin.value + dst_coin.value 
    7: MoveLoc[0](Arg0: &mut Coin<Ty0>)
    8: MutBorrowFieldGeneric[1](Coin.value: u64)
    9: WriteRef // write the result
    10: Ret
}
```
like that will be great, but unfortunately, we cant do Move Assembly.

**Comment by e99243506bigplay** [comment]
aw, I am bad on github....

**Comment by e99243506bigplay** [comment]
I hopefully I could do the optimize the merge function in bytecode level:
```
public merge_and_return_zero_coin<Ty0>(Arg0: &mut Coin<Ty0>, Arg1: Coin<Ty0>) : Coin<Ty0> {
B0:
    0: CopyLoc[0](Arg0: &mut Coin<Ty0>)
    1: ImmBorrowFieldGeneric[1](Coin.value: u64)
    2: ReadRef // stack now has [dst_coin.value]

    3: MoveLoc[1](Arg1: Coin<Ty0>) // input, stack should be [src_coin, dst_coin.value]
    4: UnpackGeneric[1](Coin<Ty0>) // unpack, stack should be [src_coin.value, dst_coin.value]

    5: Add //  dst_coin.value + src_coin.value 
    6: MoveLoc[0](Arg0: &mut Coin<Ty0>)
    7: MutBorrowFieldGeneric[1](Coin.value: u64)
    8: WriteRef // write the result
    9: LdU64(0)
    10: PackGeneric[1](Coin<Ty0>)
    11: Ret
}
```
However I unable to do that due lack of tool for now.

**Pull request closed without merging**